src/typechecker/FStar.TypeChecker.Normalize.fs

```diff
@@ -2013,15 +2013,15 @@ and norm_comp : cfg -> env -> comp -> comp =
                          | Some u -> Some <| norm_universe cfg env u
                          | None -> None
               in
-              { comp with n = Total (t, uopt) }
+              { mk_Total' t uopt with pos = comp.pos }
```

Comment:
Ah, these next few edits were fixing the evil ones. Just remarking here to remind ourselves

Building a syntax node by directly setting its `n` field is a BAD idea. It results in unintended/unsafe sharing of the free variable cache of the old node and the newly created one. This resulted in strange, hard-to-diagnose bugs which Aseem finally tracked down to this usage.

We discussed making the syntax type abstract to forbid this sort of setting of `n` directly.

We also discussed revising the nik_memoize_tc branch which uses memo tables. We can store metadata like freevars in a memo table rather than directly in the syntax node, getting rid of such issues.